openmls/src/group/public_group/process.rs

```diff
@@ -0,0 +1,234 @@
+    ///  - ValSem200
+    ///  - ValSem201
+    ///  - ValSem202: Path must be the right length
+    ///  - ValSem203: Path secrets must decrypt correctly
+    ///  - ValSem204: Public keys from Path must be verified and match the
+    ///               private keys from the direct path
+    ///  - ValSem205
+    ///  - ValSem240
+    ///  - ValSem241
+    ///  - ValSem242
+    ///  - ValSem243
+    ///  - ValSem244
+    ///  - ValSem246 (as part of ValSem010)
+    pub(crate) fn process_unverified_message(
+        &self,
+        unverified_message: UnverifiedMessage,
+        proposal_store: &ProposalStore,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<ProcessedMessage, ProcessMessageError> {
+        // Checks the following semantic validation:
+        //  - ValSem010
+        //  - ValSem246 (as part of ValSem010)
+        let (content, credential) = unverified_message.verify(backend)?;
+
+        match content.sender() {
+            Sender::Member(_) | Sender::NewMemberCommit | Sender::NewMemberProposal => {
+                let sender = content.sender().clone();
+                let authenticated_data = content.authenticated_data().to_owned();
+
+                let content = match content.content() {
+                    FramedContentBody::Application(application_message) => {
+                        ProcessedMessageContent::ApplicationMessage(ApplicationMessage::new(
+                            application_message.as_slice().to_owned(),
+                        ))
+                    }
+                    FramedContentBody::Proposal(_) => {
+                        let proposal = Box::new(QueuedProposal::from_authenticated_content(
+                            self.ciphersuite(),
+                            backend,
+                            content,
+                        )?);
+                        if matches!(sender, Sender::NewMemberProposal) {
+                            ProcessedMessageContent::ExternalJoinProposalMessage(proposal)
+                        } else {
+                            ProcessedMessageContent::ProposalMessage(proposal)
+                        }
+                    }
+                    FramedContentBody::Commit(_) => {
+                        let staged_commit = self.stage_commit(&content, proposal_store, backend)?;
+                        ProcessedMessageContent::StagedCommitMessage(Box::new(staged_commit))
+                    }
+                };
+
+                Ok(ProcessedMessage::new(
+                    self.group_id().clone(),
+                    self.group_context().epoch(),
+                    sender,
+                    authenticated_data,
+                    content,
+                    credential,
+                ))
+            }
+            Sender::External(_) => {
+                // We don't support messages from external senders yet
+                // TODO #151/#106
+                todo!()
+            }
+        }
+    }
+}
```

Comment:
Can you confirm this is just copied over from the `CoreGroup` without major changes?